gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/FlowGraphPathFinder.java

```diff
@@ -0,0 +1,428 @@
+    } catch (SpecNotFoundException | JobTemplate.TemplateException | IOException | URISyntaxException e) {
+      throw new PathFinderException(
+          "Exception encountered when computing path from src: " + srcNodeId + " to dest: " + destNodeId, e);
+    }
+  }
+
+  private boolean isPathFound(DataNode currentNode, DataNode destNode, DatasetDescriptor currentDatasetDescriptor,
+      DatasetDescriptor destDatasetDescriptor) {
+    if ((currentNode.equals(destNode)) && (currentDatasetDescriptor.equals(destDatasetDescriptor))) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * A helper method that sorts the {@link FlowEdge}s incident on srcNode based on whether the FlowEdge has an
+   * output {@link DatasetDescriptor} that is compatible with the targetDatasetDescriptor.
+   * @param dataNode
+   * @param currentDatasetDescriptor Output {@link DatasetDescriptor} of the current edge.
+   * @param destDatasetDescriptor Target {@link DatasetDescriptor}.
+   * @return prioritized list of {@link FlowEdge}s to be added to the edge queue for expansion.
+   */
+  private List<FlowEdgeContext> getNextEdges(DataNode dataNode, DatasetDescriptor currentDatasetDescriptor,
+      DatasetDescriptor destDatasetDescriptor, Config userConfig) {
+    List<FlowEdgeContext> prioritizedEdgeList = new LinkedList<>();
+    for (FlowEdge flowEdge : this.flowGraph.getEdges(dataNode)) {
+      try {
+        DataNode edgeDestination = this.flowGraph.getNode(flowEdge.getEndPoints().get(1));
+        //Should we skip this edge from any further consideration?
+        if (!edgeDestination.isActive() || !flowEdge.isActive()) {
+          continue;
+        }
+        for (Pair<DatasetDescriptor, DatasetDescriptor> datasetDescriptorPair : flowEdge.getFlowTemplate()
+            .getInputOutputDatasetDescriptors(userConfig)) {
+          DatasetDescriptor inputDatasetDescriptor = datasetDescriptorPair.getLeft();
+          DatasetDescriptor outputDatasetDescriptor = datasetDescriptorPair.getRight();
+          if (currentDatasetDescriptor.isCompatibleWith(inputDatasetDescriptor)) {
+            Config inputDatasetDescriptorConfig = inputDatasetDescriptor.getRawConfig()
+                .atPath(DatasetDescriptorConfigKeys.FLOW_EDGE_INPUT_DATASET_DESCRIPTOR_PREFIX);
+            Config outputDatasetDescriptorConfig = outputDatasetDescriptor.getRawConfig()
+                .atPath(DatasetDescriptorConfigKeys.FLOW_EDGE_OUTPUT_DATASET_DESCRIPTOR_PREFIX);
+            userConfig = userConfig.withFallback(inputDatasetDescriptorConfig).withFallback(outputDatasetDescriptorConfig);
+            if (!isFlowTemplateResolvable(flowEdge, userConfig)) {
+              continue;
+            }
+            FlowEdgeContext flowEdgeContext;
+            if (currentDatasetDescriptor.isCompatibleWith(outputDatasetDescriptor)) {
+              //If datasets described by the currentDatasetDescriptor is a subset of the datasets described
```

Comment:
Should we rename the API to isSubset(descriptor) or doesBelongsTo(descriptor)? I don't think isCompatibleWith can always infer a subset relationship, and more importantly it doesn't imply who is more specific to be preserved. What if currentDatasetDescriptor is not a subset of outputDatasetDescriptor and vice versa, but they are compatible -- then we lost outputDatasetDescriptor info?

Reply:
@yukuai518 belongsTo() is a good suggestion!

1. If currentDatasetDescriptor is not a subset of outputDatasetDescriptor (e.g. when the edge transforms input dataset from one format to another), then we choose the outputDatasetDescriptor to carry forward. That is what is being checked on line 262. Note that compatibility definition (or belongsTo()) includes both path as well as dataset format check.

2. I think the dataset descriptor should be independent of node specific information. Your suggestion to drop scheme and authority is a good one.

Reply:
Changed belongsTo() to contains().